src/overlay/TxFloodManager.cpp

```diff
@@ -0,0 +1,431 @@
+        demandMap;
+    bool anyNewDemand = false;
+    do
+    {
+        anyNewDemand = false;
+        for (auto const& peer : peers)
+        {
+            auto& demPair = demandMap[peer];
+            auto& demand = demPair.first;
+            auto& retry = demPair.second;
+            bool addedNewDemand = false;
+
+            auto it = mQueuedIncomingAdverts.find(peer);
+            if (it == mQueuedIncomingAdverts.end())
+            {
+                continue;
+            }
+            auto& queue = *(it->second);
+            while (demand.size() < getMaxDemandSize() && queue.size() > 0 &&
+                   !addedNewDemand)
+            {
+                auto hashPair = queue.pop();
+                auto txHash = hashPair.first;
+                if (hashPair.second)
+                {
+                    auto delta = now - *(hashPair.second);
+                    om.mAdvertQueueDelay.Update(delta);
+                    peer->getPeerMetrics().mAdvertQueueDelay.Update(delta);
+                }
+                switch (demandStatus(txHash, peer))
+                {
+                case DemandStatus::DEMAND:
+                    demand.push_back(txHash);
+                    if (mDemandHistoryMap.find(txHash) ==
+                        mDemandHistoryMap.end())
+                    {
+                        // We don't have any pending demand record of this tx
+                        // hash.
+                        mPendingDemands.push(txHash);
+                        mDemandHistoryMap[txHash].firstDemanded = now;
+                        CLOG_DEBUG(Overlay, "Demand tx {}, asking peer {}",
+                                   hexAbbrev(txHash), peer->toString());
+                    }
+                    else
+                    {
+                        om.mDemandTimeouts.Mark();
+                        ++(peer->getPeerMetrics().mDemandTimeouts);
+                    }
+                    mDemandHistoryMap[txHash].peers.emplace(peer->getPeerID(),
+                                                            now);
+                    mDemandHistoryMap[txHash].lastDemanded = now;
+                    addedNewDemand = true;
+                    break;
+                case DemandStatus::RETRY_LATER:
+                    retry.push_back(txHash);
+                    break;
+                case DemandStatus::DISCARD:
+                    break;
+                }
+            }
+            anyNewDemand |= addedNewDemand;
+        }
+    } while (anyNewDemand);
+
+    for (auto const& peer : peers)
+    {
+        // We move `demand` here and also pass `retry` as a reference
+        // which gets appended. Don't touch `demand` or `retry` after here.
+        peer->sendTxDemand(std::move(demandMap[peer].first));
+        if (demandMap[peer].second.empty())
+        {
+            continue;
+        }
+        auto it = mQueuedIncomingAdverts.find(peer);
+        if (it == mQueuedIncomingAdverts.end())
```

Comment:
TODO: peer must already be present in incoming adverts, otherwise can't populate retries to begin with.